pageserver/src/tenant/timeline/init.rs

```diff
@@ -39,14 +40,27 @@ pub(super) enum Discovered {
 pub(super) fn scan_timeline_dir(path: &Utf8Path) -> anyhow::Result<Vec<Discovered>> {
     let mut ret = Vec::new();
 
+    let gen_suffix_regex = Regex::new("^(?<base>.+)-(?<gen>[0-9a-f]{8})$").unwrap();
+
     for direntry in path.read_dir_utf8()? {
         let direntry = direntry?;
         let file_name = direntry.file_name().to_string();
 
+        // If the local path has a generation suffix, trim that off before trying to parse
+        // it as a LayerFileName.
+        let file_name = match gen_suffix_regex.captures(&file_name) {
+            Some(captures) => captures
+                .name("base")
+                .expect("Non-optional group")
+                .as_str()
+                .to_owned(),
+            None => file_name,
+        };
```

Comment:
This code is the same as in `LayerFileName::from_str`.

I already complained about the `from_str` impl not feeling super idiomatic.

Maybe we can have a `struct LocalLayerFileName { base: LayerFileName, gen: Generation }` and move the `FromStr` impl there?

Reply:
Cleaned up the duplication in 0a2606e1c